addons/mail/static/src/messaging/component/message_list/message_list.js

```diff
@@ -283,85 +276,98 @@ class MessageList extends Component {
         }
-        const {
-            isLastMessageVisible,
-            isPatchedWithLoadMoreMessages,
-            isPatchedWithNewMessages,
-            isPatchedWithNewThreadCache,
-            scrollHeight: beforeScrollHeight,
-            scrollTop: beforeScrollTop,
-        } = this._willPatchSnapshot;
-        if (isPatchedWithLoadMoreMessages) {
-            if (this.props.order === 'asc') {
-                this.el.scrollTop = this.el.scrollHeight - beforeScrollHeight + beforeScrollTop;
+    }
+
+    /**
+     * @private
+     * @param {Object} hint
+     * @param {integer} hint.messageId
+     */
+    async _adjustFromCurrentPartnerJustPostedMessage(hint) {
+        const threadCache = this.threadViewer.threadCache;
+        const { messageId } = hint;
+        if (threadCache.isLoaded) {
+            const threadCacheMessageIds = threadCache.messages.map(message => message.id);
+            if (threadCacheMessageIds.includes(messageId) && this.messageRefFromId(messageId)) {
+                await this._scrollToMessage(messageId);
+                this.threadViewer.markComponentHintProcessed(hint);
             }
-        } else if (isPatchedWithNewThreadCache || (isPatchedWithNewMessages && isLastMessageVisible)) {
-            this.scrollToLastMessage().then(() => this._onScroll());
-        } else if (
-            isPatchedWithNewMessages &&
-            this.storeProps.threadCacheCurrentPartnerMessagePostCounter !== this._threadCacheCurrentPartnerMessagePostCounter
-        ) {
-            this._scrollToLastCurrentPartnerMessage().then(() => this._onScroll());
-        } else if (
-            !this._selectedMessageLocalId &&
-            this._selectedMessageLocalId !== this.props.selectedMessageLocalId &&
-            !this.storeProps.isMobile
-        ) {
-            this._scrollToMessage(this.props.selectedMessageLocalId).then(() => this._onScroll());
         }
-        this._willPatchSnapshot = undefined;
     }
 
-    //--------------------------------------------------------------------------
-    // Private
-    //--------------------------------------------------------------------------
+    /**
+     * @private
+     * @param {Object} hint
+     */
+    _adjustFromMoreMessagesLoaded(hint) {
+        if (!this._willPatchSnapshot) {
+            this.threadViewer.markComponentHintProcessed(hint);
+            return;
+        }
+        const { scrollHeight, scrollTop } = this._willPatchSnapshot;
+        if (this.props.order === 'asc') {
+            this.el.scrollTop = this.el.scrollHeight - scrollHeight + scrollTop;
+        }
+        this.threadViewer.markComponentHintProcessed(hint);
+    }
 
     /**
      * @private
      */
     _checkThreadMarkAsRead() {
-        if (this.storeProps.messages.length === 0) {
+        const thread = this.threadViewer.thread;
+        const threadCache = this.threadViewer.threadCache;
+        if (!threadCache) {
+            // FIXME: component uses obsolete store data...
```

Comment:
fixme remaining (if that is part of all of the "if" that are necessary for the same reason, we might not want to comment it, or if we do, then do it everywhere).